### PR TITLE
feat: Add Bootstrap 4 Theme support

### LINKED ISSUE #1455: Support bootstrap 4
Could someone make a theme for bootstrap 4 using `withTheme`? That would be great 😃

### PR DESCRIPTION
**Pull request by Xtremilicious**
### Reasons for making this change

To add Bootstrap 4 theme support for react-jsonschema-form as a task for the MLH Fellowship 2020.

Fix #1455 

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [x] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
* [x] **I'm adding a new feature**
  - [x] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/bootstrap-4/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+<!-- PROJECT LOGO -->
+<br />
+<p align="center">
+  <a href="https://github.com/cybertec-postgresql/rjsf-material-ui">
```
Comment (epicfaace) [inline packages/bootstrap-4/README.md:4]: Change this link

File: packages/bootstrap-4/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import { Theme as Bootstrap4Theme } from '@rjsf/bootstrap-4';
+
+const Form = withTheme(Bootstrap4Theme);
+```
+
+<!-- MARKDOWN LINKS & IMAGES -->
+<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
+
+[build-shield]: https://img.shields.io/circleci/build/github/cybertec-postgresql/rjsf-material-ui.svg?style=flat-square&token=a58b0890f96bff2b53eef0f4d9c9e5d16eec2200
+[build-url]: https://circleci.com/gh/cybertec-postgresql/rjsf-material-ui
```
Comment (epicfaace) [inline packages/bootstrap-4/README.md:75]: Change these links too -- take a look at one of the README.md's for the existing packages to see what you could put instead.

File: packages/bootstrap-4/test/addButton.test.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
```
Comment (epicfaace) [inline packages/bootstrap-4/test/addButton.test.tsx:1]: Can you capitalize all the file names of tests (since they're components too, for consistency)?

File: packages/bootstrap-4/test/arrayFieldTemplate.test.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+          onAddClick={() => void 0}
+          readonly={false}
+          required={false}
+          schema={mockSchema}
+          uiSchema={{}}
+          title=""
+          formContext={{}}
+          formData={{}}
+          registry={{ ...getDefaultRegistry() }}
+          // TODO : isSchema should be fixed here
```
Comment (epicfaace) [inline packages/bootstrap-4/test/arrayFieldTemplate.test.tsx:29]: What is this for?

File: packages/core/index.d.ts [hunk 5]
```diff
@@ -173,7 +173,7 @@ declare module '@rjsf/core' {
         className: string;
         disabled: boolean;
         idSchema: IdSchema;
-        items: {
+        items?: {
```
Comment (epicfaace) [inline packages/core/index.d.ts:176]: Why should this be optional?

File: packages/playground/package.json [hunk 6]
```diff
@@ -58,6 +58,9 @@
     "prop-types": "^15.7.2",
     "react-app-polyfill": "^1.0.4",
     "react-bootstrap": "^0.33.1",
+    "@fortawesome/free-solid-svg-icons": "^5.13.1",
```
Comment (epicfaace) [inline packages/playground/package.json]: Why are these needed? Should we include these as dependencies / peerDependencies for @rjsf/bootstrap-4 and mention them in the readme?

File: packages/playground/src/index.js [hunk 7]
```diff
@@ -3,6 +3,7 @@ import { Theme as MuiTheme } from "@rjsf/material-ui";
 import { Theme as FluentUITheme } from "@rjsf/fluent-ui";
 import { Theme as SuiTheme } from "@rjsf/semantic-ui";
 import { Theme as AntdTheme } from "@rjsf/antd";
+import { Theme as Bootstrap4Theme } from "../../bootstrap-4";
```
Comment (epicfaace) [inline packages/playground/src/index.js:6]:
```suggestion
import { Theme as Bootstrap4Theme } from "@rjsf/bootstrap-4";
```

File: packages/playground/package.json (hunk 6 above)
Comment (anikethsaha) [reply packages/playground/package.json]: we removed this just now. it should be fixed in the latest commit

File: packages/playground/src/index.js (hunk 7 above)
Comment (anikethsaha) [reply packages/playground/src/index.js:6]: Like as of now it is not published right, so for the CI to work fine. we are using this. Once CI is passed, we can change it in this PR or in a separate PR

File: packages/core/index.d.ts (hunk 5 above)
Comment (anikethsaha) [reply packages/core/index.d.ts:176]: may be we missed something here, let me check

File: packages/bootstrap-4/test/arrayFieldTemplate.test.tsx (hunk 4 above)
Comment (anikethsaha) [reply packages/bootstrap-4/test/arrayFieldTemplate.test.tsx:29]: actually we are not able to create  a mock entry for `idSchema` as its bit complicated. 
So we skipped it. 

Can you suggest a mock entry for that so we I can undo that and I will add that.

File: packages/playground/src/index.js (hunk 7 above)
Comment (epicfaace) [reply packages/playground/src/index.js:6]: If you run `npm install` from the root directory, lerna will symlink `@rjsf/bootstrap-4` to point to the local directory. It doesn't need to be published.

**Comment by adrianparvino** [comment]
This branch does not seem to include the ObjectFieldTemplate patch
https://github.com/MLH-Fellowship/react-jsonschema-form/blob/react-bootstrap-4-final/packages/bootstrap-4/src/ObjectFieldTemplate/ObjectFieldTemplate.tsx

from
https://github.com/MLH-Fellowship/react-jsonschema-form/blob/react-bootstrap-4/packages/bootstrap-4/src/ObjectFieldTemplate/ObjectFieldTemplate.tsx

**Comment by anikethsaha** [comment]
Thanks a lot for pointing that, we might have missed that while rebasing

**Comment by Xtremilicious** [comment]
@epicfaace the suggestions mentioned have been implemented. Kindly have a look :)

File: packages/playground/src/index.js (hunk 7 above)
Comment (Xtremilicious) [reply packages/playground/src/index.js]: updated

**Comment by epicfaace** [comment]
Some comments after looking through the playground:

I think the description field is too big here ("Simple" tab):

image


Any idea why the "nested" example has cut off text?

image


"nested" example -- can you make sure the buttons ("+" and "-") have a consistent size?

image

"nested" example -- the "Done?" is supposed to be a checkbox, but it's rendering as a big title:

image

"nested" example -- apparently `[object Object]` are being passed in for the uiSchema and registry HTML attributes for the `<input>` tag. Can you remove them?

image

"widgets" example -- the read-only fields should have the `readonly` attribute set, not the `disabled` attribute.

image

"simple" example -- whenever I type anything in a textbox, I get this error in the console:

image

Validation - the validation doesn't seem to look quite like what default validation in bootstrap looks like. Can you use this to render inline validation errors instead: https://react-bootstrap.github.io/components/forms/#forms-validation-libraries

image

Error List - any idea why the bottom left and bottom right corners aren't connected? Also, can you add a bit more margin below the error list (like how it looks like for the bootstrap 3 form)?

image

"Widgets" tab - I am not able to click on "No" to change the value for the radio button.

image


"Files" tab - can you change the file input to use Bootstrap's file input (https://getbootstrap.com/docs/4.0/components/forms/#file-browser)?

image

"Alternatives" tab - I can't select anything in the multiselect widget:

image

"Additional Properties" tab - can you implement the additional properties functionality here?

image

"Nested" tab - can you make the help text grey with the `text-muted` class as described in the [Bootstrap docs](https://getbootstrap.com/docs/4.0/components/forms/#help-text)? Additionally, 

image

**Comment by anikethsaha** [comment]
@epicfaace We have updated the tests and applied suggestions. 
CI tests are passing. Not sure about the netlify thing.

**Comment by epicfaace** [comment]
@anikethsaha can you merge the latest master into this branch? That should fix the netlify issue.

**Comment by Xtremilicious** [comment]
> @anikethsaha can you merge the latest master into this branch? That should fix the netlify issue.

Did the merge, still failing. Log shows some problem with `sh: cross-env: command not found`?

**Comment by anikethsaha** [comment]
@epicfaace it seems the build is still failing.

From the logs, it looks like it is simply running the `build` command instead of installing the dependencies I think. 

let us know if anything else is required to fix from our end

File: packages/bootstrap-4/src/DescriptionField/DescriptionField.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import React from "react";
+import { FieldProps } from "@rjsf/core";
+
+export interface DescriptionFieldProps extends Partial<FieldProps> {
+  description?: string;
+}
+
+const DescriptionField = ({ description }: Partial<FieldProps>) => {
+  if (description) {
+    return <div><h6 className="mb-5">{description}</h6></div>;
```
Comment (epicfaace) [inline packages/bootstrap-4/src/DescriptionField/DescriptionField.tsx:10]: We don't want descriptions to be in bold

```suggestion
    return <div><div className="mb-5">{description}</div></div>;
```

File: packages/bootstrap-4/src/ErrorList/ErrorList.tsx [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+const ErrorList = ({ errors }: ErrorListProps) => (
+  <Card border="danger" className="mb-2">
+    <Card.Header className="alert-danger">Errors</Card.Header>
+    <Card.Body className="p-0">
+      <ListGroup >
+        {errors.map((error, i: number) => {
+          return (
+            <ListGroup.Item key={i} className="border-0">
+              <small>{error.stack}</small>
```
Comment (epicfaace) [inline packages/bootstrap-4/src/ErrorList/ErrorList.tsx:16]: Let's use regular-sized text for ErrorList

```suggestion
              <span>{error.stack}</span>
```

File: packages/bootstrap-4/src/ErrorList/ErrorList.tsx [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+
+import Card from "react-bootstrap/Card";
+import ListGroup from "react-bootstrap/ListGroup";
+
+import { ErrorListProps } from "@rjsf/core";
+
+const ErrorList = ({ errors }: ErrorListProps) => (
+  <Card border="danger" className="mb-2">
```
Comment (epicfaace) [inline packages/bootstrap-4/src/ErrorList/ErrorList.tsx:9]: Need more margin between this and the form title

```suggestion
  <Card border="danger" className="mb-4">
```

**Comment by anikethsaha** [comment]
@epicfaace I have updated the code with your changes. 

Let us know if there is anything else required.

**Comment by nicmendoza** [comment]
I'm super excited to try this out -- are you planning on updating the documentation so it references this theme? I'm not clear how to actually use it. It appears, based on the other themes, that I need to install @rjsf/bootstrap-4, which doesn't exist.

**Comment by epicfaace** [comment]
Yeah, it isn't released to npm yet -- it will be released on the next release.

@anikethsaha @Xtremilicious could you make a PR so that it updates the theme list table on the documentation to add this theme?

**Comment by anikethsaha** [comment]
you meant here https://github.com/rjsf-team/react-jsonschema-form#supported-themes ?

**Comment by epicfaace** [comment]
Yes, that link, and this link: https://react-jsonschema-form.readthedocs.io/en/latest/usage/themes/#supported-themes

**Comment by epicfaace** [comment]
[@rjsf/bootstrap-4](https://www.npmjs.com/package/@rjsf/bootstrap-4) has just been published to npm with version 2.3.0. Thanks for your patience!
